### ISSUE TITLE
Writing to project.lock.json fails sometimes when creating a web project

### ISSUE THREAD

**Issue by barrytang** [dotnet/cli#1762]
## Steps to reproduce
Create a web project

## Expected  behavior
No error in the Error list

## Environment data
`dotnet --version` output:

**Comment by barrytang** [comment]
![image](https://cloud.githubusercontent.com/assets/1648865/13617843/9380824e-e535-11e5-8f78-861ac96b4e83.png)

**Comment by BillHiebert** [comment]
I am seeing builds fail with this same issue, and this is affecting a lot of our tests..

**Comment by piotrpMSFT** [comment]
@pakrym, I just spoke with @BillHiebert and verified that VS itself is not likely to be looking at project.lock.json. There is suspicion that ProjectModelServer might be keeping the file in use. Any thoughts?

**Comment by troydai** [comment]
[`WorkspaceContext`](https://github.com/dotnet/cli/blob/rel/1.0.0/src/Microsoft.DotNet.ProjectModel/WorkspaceContext.cs) is where we cached the `ProjectContext` and `LockFile`. `WorkspaceContext` itself doesn't cache a file handler by itself. So I'm looking at `LockFile` and `ProjectContext` now.

**Comment by barrytang** [comment]
+@Eilon

@troydai, any updates? This is blocking us from running tests reliably...

**Comment by troydai** [comment]
Sorry I was sidetracked this morning. I'll run a simulation now.

**Comment by troydai** [comment]
I doubt it is caused by project model server. In this [test](https://github.com/dotnet/cli/blob/troy/1668/test/dotnet-projectmodel-server.Tests/DthTests.cs#L392) the lockfile is updated while server is still running and `WorkspaceContext` is still catching all the data. The test pass on both Windows and Mac OSX.

**Comment by troydai** [comment]
Todd and I were trying to repo this just now. We created a project, wait for the restore and delete the lock file. No error. The lock file is deleted successful. That means dotnet-projectmodel-server doesn't occupy the lock file. 

More like it is related to the VS started `dotnet restore` commands. Could there be parallel `restore` happen?

**Comment by troydai** [comment]
@barrytang

**Comment by troydai** [comment]
Update: Todd manage to create a repo and save the Process Monitor logs. I'm investigating.

**Comment by troydai** [comment]
```
3:33:22.9711109 PM	dotnet.exe	21896	CreateFile	C:\Users\toddgrun\Desktop\webs\EmptyK\project.lock.json	SHARING VIOLATION	Desired Access: Generic Read, Disposition: Open, Options: Non-Directory File, Open No Recall, Attributes: n/a, ShareMode: Read, AllocationSize: n/a
```
This shows up in the log and highly suspicious. What interesting is that the `Generic Read` is only desired on one of the 2 lock file.

**Comment by troydai** [comment]
The issue is other way around. `dotnet restore` is occupying the lock file and in the meantime pm server tries to read lock file. We need a retry mechanism to deal with this situation.

Assign this to myself.

**Comment by troydai** [comment]
Fixed in 37544303cbdc1a545b3113a5d2c2d81a1cc9492e
